azimuth/utils/ml/perturbation_functions.py

```diff
@@ -307,9 +309,19 @@ def typo(original: str, config: PerturbationTestingConfig) -> List[PerturbedUtte
         with contextlib.redirect_stdout(None):
             # While nlpaug fixes their useless print, we ignore it.
             perturbed_utterance = aug.augment(original, n=1)
-        # Issue 854 nac adds spaces between the hyphen and $ sign.
-        perturbed_utterance = perturbed_utterance.replace(" ' ", "'")
-        perturbed_utterance = perturbed_utterance.replace("$ ", "$")
+        # Issue 854 nac adds spaces around the quote/apostrophe and $ sign; remove them.
+        perturbed_utterance = re.sub(r" (['’]) ", r"\1", perturbed_utterance)
+        perturbed_utterance = re.sub(r"([$]) ", r"\1", perturbed_utterance)
```

Comment:
You might have noticed that `$` is a special character in regular expressions. Inside a `[]`, special characters are not considered special (except `-`), so your regexp works. Another way to tell that you want a literal `$` is to escape it with a `\`.
```suggestion
        perturbed_utterance = re.sub(r"(\$) ", r"\1", perturbed_utterance)
```
By the way, if you're interested in regular expressions, here is another way to approach the problem:
We can match only the space, on the condition that it is preceded with a `$` (but not matching it), with a lookbehind non-capturing group. That way, we don't need to use a capturing group (which takes a little bit of memory and time), we simply replace the space with an empty string.
```suggestion
        perturbed_utterance = re.sub(r"(?<=\$) ", "", perturbed_utterance)
```
In this situation, however, I don't think it makes a difference. It's just cool to know.

Reply:
(Sorry, unedited comment had a bunch of ?<->$ confusion because I was thinking about the other regex!)

Reply:
Correct. As soon as there is a `?` after the `(`, it's not a capturing group. I emphasized on the fact it was non-capturing to explain why it is more efficient to use it.
Your suggestion wouldn't work as it would remove the `$`. The `(?:)` is non-capturing, but it is still matching, whereas a lookbehind or lookahead doesn't include the `$` in the match. You can think of capturing as saving some "sub-matches" in memory to reuse it later, like with your `\1` replacement. That's what is a tiny bit more expensive ($, expensive, lol). Makes sense?
The basic non-capturing group is useful for making a string (more than one character) optional, without wasting memory to capture the group. For example,
```
(?:ex)?press
```
would match `express` or `press`, but not `xpress` nor `epress` (whereas the regexp `e?x?press` would)